src/compiler/checker.ts

```diff
@@ -10123,6 +10170,37 @@ namespace ts {
             return resolveErrorCall(node);
         }
 
+        function isConstructorAccessible(node: NewExpression, signature: Signature) {
+            if (!signature || !signature.declaration) {
+                return true;
+            }
+
+            const declaration = signature.declaration;
+            const flags = declaration.flags;
+
+            // Public constructor is accessible.
+            if (!(flags & (NodeFlags.Private | NodeFlags.Protected))) {
+                return true;
+            }
+
+            const declaringClass = <InterfaceType>getDeclaredTypeOfSymbol(declaration.parent.symbol);
+            const enclosingClassDeclaration = getContainingClass(node);
+            const enclosingClass = enclosingClassDeclaration ? <InterfaceType>getDeclaredTypeOfSymbol(getSymbolOfNode(enclosingClassDeclaration)) : undefined;
+
+            // A private or protected constructor can only be instantiated within it's own class 
+            if (declaringClass !== enclosingClass) {

```

Comment:
as noted in , we https://github.com/Microsoft/TypeScript/issues/7059, if here is not sufficient, it should be a while loop walking up, looking at all enclosing classes, so that something like this should be legal:

```ts

class B {
    private constructor() { }

    method() {
        class C {
            method() {
                new B(); // should be fine
            }
        }
    }
}
```